src/Microsoft.OData.Client/Serialization/Serializer.cs

```diff
@@ -405,6 +614,139 @@ internal void WriteNestedResourceInfo(EntityDescriptor entityDescriptor, IEnumer
+                            if (relatedLinks == null)
+                            {
+                                relatedLinks = new List<Descriptor>();
+                            }
+
+                            relatedLinks.Add(entityObj);
+                            groupRelatedLinks.Add(parentProperty, relatedLinks);
+                        }
+                  
+                        break;
+                    case DescriptorKind.Link:
+                        var linkObj = (LinkDescriptor)link;
+                        var linkName = linkObj.SourceProperty;
+                        if (groupRelatedLinks.TryGetValue(linkName, out relatedLinks))
+                        {
+                            relatedLinks.Add(linkObj);
+                        }
+                        else
+                        {
+                            if (relatedLinks == null)
+                            {
+                                relatedLinks = new List<Descriptor>();
+                            }
+
+                            relatedLinks.Add(linkObj);
+                            groupRelatedLinks.Add(linkName, relatedLinks);
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            ClientTypeAnnotation clientType = null;
+            if (clientType == null)
+            {
+                ClientEdmModel model = this.requestInfo.Model;
+                clientType = model.GetClientTypeAnnotation(model.GetOrCreateEdmType(entityDescriptor.Entity.GetType()));
+            }
+
+            
+
+            foreach (KeyValuePair<string, List<Descriptor>> relatedLinks in groupRelatedLinks)
+            {
+                bool isCollection = false;
+                bool started = false;
+
+                LinkDescriptor relatedEnd = null;
+                ODataNestedResourceInfo navigationLink = new ODataNestedResourceInfo();
+                
+                //these items will all be of the same type ~ we'll be using the 
+                //first item in determining the type of the descriptor.
+                Descriptor item = relatedLinks.Value.FirstOrDefault();
+    
+                if (item.DescriptorKind == DescriptorKind.Entity)
+                {
+                    var entDescriptor = item as EntityDescriptor;
+
+                    if (entDescriptor.IsDeepInsert)
+                    {
+                        relatedEnd = entDescriptor.GetRelatedEnd();
+                    }
+                    else
+                    {
+                        relatedEnd = new LinkDescriptor(entDescriptor.ParentEntity, entDescriptor.ParentProperty, entDescriptor.Entity, this.requestInfo.Model);
+                    }
+                }
+                else
+                {
+                    relatedEnd = item as LinkDescriptor;
+                }
+
+                navigationLink.Name = relatedEnd.SourceProperty;

```

Comment:
is this the same thing as `relatedLinks.Key`?

Reply:
No..`relatedEnd` is a `LinkDescriptor` that has a property called `SourceProperty.`